pkg/controller/v1beta1/inferenceservice/utils/utils.go

```diff
@@ -58,3 +67,99 @@ func GetDeploymentMode(annotations map[string]string, deployConfig *v1beta1api.D
 	}
 	return constants.DeploymentModeType(deployConfig.DefaultDeploymentMode)
 }
+
+// Merge the predictor Container struct with the runtime Container struct, allowing users
+// to override runtime container settings from the predictor spec.
+func MergeRuntimeContainers(runtimeContainer *v1alpha1.Container, predictorContainer *v1.Container) (*v1.Container, error) {
+	// Default container configuration from the runtime.
+	coreContainer := v1.Container{
+		Args:            runtimeContainer.Args,
+		Command:         runtimeContainer.Command,
+		Env:             runtimeContainer.Env,
+		Image:           runtimeContainer.Image,
+		Name:            runtimeContainer.Name,
+		Resources:       runtimeContainer.Resources,
+		ImagePullPolicy: runtimeContainer.ImagePullPolicy,
+		WorkingDir:      runtimeContainer.WorkingDir,
+		LivenessProbe:   runtimeContainer.LivenessProbe,
+	}
+	// Save runtime container name, as the name can be overridden as empty string during the Unmarshal below
+	// since the Name field does not have the 'omitempty' struct tag.
+	runtimeContainerName := runtimeContainer.Name
+
+	// Args and Env will be combined instead of overridden.
+	argCopy := make([]string, len(coreContainer.Args))
+	copy(argCopy, coreContainer.Args)
+
+	envCopy := make([]v1.EnvVar, len(coreContainer.Env))
+	copy(envCopy, coreContainer.Env)
+
+	// Use JSON Marshal/Unmarshal to merge Container structs.
+	overrides, err := json.Marshal(predictorContainer)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := json.Unmarshal(overrides, &coreContainer); err != nil {
+		return nil, err
+	}
+
+	if coreContainer.Name == "" {
+		coreContainer.Name = runtimeContainerName
+	}
+
+	argCopy = append(argCopy, predictorContainer.Args...)
+	envCopy = append(envCopy, predictorContainer.Env...)
+
+	coreContainer.Args = argCopy
+	coreContainer.Env = envCopy
+
+	return &coreContainer, nil
+
+}
+
+// Merge the predictor PodSpec struct with the runtime PodSpec struct, allowing users
+// to override runtime PodSpec settings from the predictor spec.
+func MergePodSpec(runtimePodSpec *v1alpha1.ServingRuntimePodSpec, predictorPodSpec *v1beta1.PodSpec) (*v1.PodSpec, error) {
```

Comment:
Were you thinking we'll end up with a single PodSpec definition and if so which one? I'm trying to remember why we went with more restricted versions of the Pod and Container specs, there was some reason I think that we wanted to be able to not allow _full_ control.

Reply:
Right, for now just trying to allow the level of configuration that KServe currently provides since the Isvc predictor spec has the full podspec as available fields. PodSpec fields in the Isvc will override podspec fields in the SR. I realize a lot of this is more single model serving centric, as in the modelmesh case, I suppose PodSpec fields specified in the Isvc won't have any effect.